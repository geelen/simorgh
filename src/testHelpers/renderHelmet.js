/*
 * Similar to this problem https://github.com/testing-library/react-testing-library/issues/402
 * This helper was created to solve the problem of rendering helmet/head contents in snapshots.
 * Pass in a component that uses helmet somewhere in the component tree.
 * The full html tree is rendered and returned which you can then use to snapshot helmet/head contents.
 * Example of use in a test file:

 it('should render correctly', async () => {
    const html = await renderHelmet(<SomeComponent {...someProps} />);
    expect(html).toMatchSnapshot();
  });

 */

import { render, waitForDomChange } from '@testing-library/react';

export default async component => {
  render(component);

  await waitForDomChange();

  const htmlElement = document.querySelector('html');

  const helmetElements = document.querySelectorAll(
    '[data-react-helmet="true"]',
  );

  const removeHelmetAttributes = el => el.removeAttribute('data-react-helmet'); // remove react-helmet attribute noise from elements

  removeHelmetAttributes(htmlElement); // remove react-helmet attribute noise from elements

  Array.from(helmetElements).forEach(removeHelmetAttributes);

  return document.querySelector('html');
};
