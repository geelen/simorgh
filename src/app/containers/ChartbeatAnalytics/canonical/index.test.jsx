import React from 'react';
import { mount } from 'enzyme';
import renderHelmet from '../../../../testHelpers/renderHelmet';
import CanonicalChartbeatAnalytics from '.';

describe('CanonicalChartbeatAnalytics', () => {
  global.pSUPERFLY = {
    virtualPage: jest.fn(),
  };

  const config = {
    domain: 'test-domain',
    type: 'article',
    sections: 'section1 section2',
    chartbeatUID: 1111,
    virtualReferrer: null,
    useCanonical: true,
    title: 'This is an article',
    uid: 123,
  };

  it('should return the helmet wrapper with the script snippet', async () => {
    const html = await renderHelmet(
      <CanonicalChartbeatAnalytics
        chartbeatConfig={config}
        chartbeatSource="//chartbeat.js"
      />,
    );

    expect(html).toMatchSnapshot();
  });

  it('should call the global virtualPage function when props change', () => {
    const wrapper = mount(
      <CanonicalChartbeatAnalytics
        chartbeatConfig={config}
        chartbeatSource="//chartbeat.js"
      />,
    );

    wrapper.setProps({
      chartbeatConfig: {
        domain: 'test-domain',
        type: 'article',
        sections: 'section1 section2',
        chartbeatUID: 1111,
        useCanonical: true,
        virtualReferrer: '/some-path',
        title: 'This is another article',
        uid: 123,
      },
    });
    wrapper.unmount();
    expect(global.pSUPERFLY.virtualPage).toHaveBeenCalled();
    expect(global.pSUPERFLY.virtualPage).toHaveBeenCalledWith({
      domain: 'test-domain',
      type: 'article',
      sections: 'section1 section2',
      chartbeatUID: 1111,
      useCanonical: true,
      virtualReferrer: '/some-path',
      title: 'This is another article',
      uid: 123,
    });
  });
});
