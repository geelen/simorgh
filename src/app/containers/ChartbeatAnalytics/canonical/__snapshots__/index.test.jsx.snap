// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanonicalChartbeatAnalytics should return the helmet wrapper with the script snippet 1`] = `
<html>
  <head>
    <script
      async="true"
      type="text/javascript"
    >
      
        (function(){
          var _sf_async_config = window._sf_async_config = (window._sf_async_config || {});
          var config = {"domain":"test-domain","type":"article","sections":"section1 section2","chartbeatUID":1111,"virtualReferrer":null,"useCanonical":true,"title":"This is an article","uid":123};
          for (var key in config) {
            _sf_async_config[key] = config[key];
          }
        })();
      
    </script>
    <script
      async="true"
      src="//chartbeat.js"
      type="text/javascript"
    />
  </head>
  <body>
    <div />
  </body>
</html>
`;
